File: algencan-3.1.1/sources/examples/f77/toyprob2.f
C     ******************************************************************
C     ******************************************************************

      program algencanma

      implicit none

C     PARAMETERS
      integer nmax,mmax
      parameter ( nmax = 2 )
      parameter ( mmax = 2 )

C     LOCAL SCALARS
      logical checkder
      integer hnnzmax,hnnzmax12,hnnzmax3,inform,jcnnzmax,m,n,nvparam
      double precision cnorm,efacc,efstain,eoacc,epsfeas,epsopt,eostain,
     +        f,nlpsupn,snorm

C     LOCAL ARRAYS
      character * 80 specfnm,outputfnm,vparam(10)
      logical coded(11),equatn(mmax),linear(mmax)
      double precision l(nmax),lambda(mmax),u(nmax),x(nmax)

C     EXTERNAL SUBROUTINES
      external myevalf,myevalg,myevalh,myevalc,myevaljac,myevalhc,
     +         myevalfc,myevalgjac,myevalgjacp,myevalhl,myevalhlp

C     Number of variables

      n = 2

C     Set lower bounds, upper bounds, and initial guess

      l(1) = - 10.0d0
      u(1) =   10.0d0

      l(2) = - 1.0d+20
      u(2) =   1.0d+20

      x(1) =   0.0d0
      x(2) =   0.0d0

C     Constraints

      m = 2

      equatn(1) = .false.
      equatn(2) = .false.

      lambda(1) = 0.0d0
      lambda(2) = 0.0d0

      linear(1) = .false.
      linear(2) = .true.

C     Coded subroutines

      coded( 1) = .false. ! fsub
      coded( 2) = .false. ! gsub
      coded( 3) = .false. ! hsub
      coded( 4) = .false. ! csub
      coded( 5) = .false. ! jacsub
      coded( 6) = .false. ! hcsub
      coded( 7) = .true.  ! fcsub
      coded( 8) = .true.  ! gjacsub
      coded( 9) = .false. ! gjacpsub
      coded(10) = .true.  ! hlsub
      coded(11) = .false. ! hlpsub

C     Upper bounds on the number of sparse-matrices non-null elements

      jcnnzmax  = 4
      hnnzmax12 = 1
      hnnzmax3  = 6
      hnnzmax   = hnnzmax12 + hnnzmax3

C     Checking derivatives?

      checkder = .true.

C     Parameters setting

      epsfeas   = 1.0d-08
      epsopt    = 1.0d-08

      efstain   = sqrt( epsfeas )
      eostain   = epsopt ** 1.5d0

      efacc     = sqrt( epsfeas )
      eoacc     = sqrt( epsopt )

      outputfnm = ''
      specfnm   = ''

      nvparam = 0

      call algencan(myevalf,myevalg,myevalh,myevalc,myevaljac,myevalhc,
     +myevalfc,myevalgjac,myevalgjacp,myevalhl,myevalhlp,jcnnzmax,
     +hnnzmax,epsfeas,epsopt,efstain,eostain,efacc,eoacc,outputfnm,
     +specfnm,nvparam,vparam,n,x,l,u,m,lambda,equatn,linear,coded,
     +checkder,f,cnorm,snorm,nlpsupn,inform)

      stop

      end

C     ******************************************************************
C     ******************************************************************

      subroutine myevalf(n,x,f,flag)

      implicit none

C     SCALAR ARGUMENTS
      integer flag,n
      double precision f

C     ARRAY ARGUMENTS
      double precision x(n)

      flag = - 1

      end

C     ******************************************************************
C     ******************************************************************

      subroutine myevalg(n,x,g,flag)

      implicit none

C     SCALAR ARGUMENTS
      integer flag,n

C     ARRAY ARGUMENTS
      double precision g(n),x(n)

      flag = - 1

      end

C     ******************************************************************
C     ******************************************************************

      subroutine myevalh(n,x,hrow,hcol,hval,hnnz,lim,lmem,flag)

      implicit none

C     SCALAR ARGUMENTS
      logical lmem
      integer flag,n,hnnz,lim

C     ARRAY ARGUMENTS
      integer hcol(lim),hrow(lim)
      double precision hval(lim),x(n)

      flag = - 1

      end

C     ******************************************************************
C     ******************************************************************

      subroutine myevalc(n,x,ind,c,flag)

      implicit none

C     SCALAR ARGUMENTS
      integer ind,flag,n
      double precision c

C     ARRAY ARGUMENTS
      double precision x(n)

      flag = - 1

      end

C     ******************************************************************
C     ******************************************************************

      subroutine myevaljac(n,x,ind,jcvar,jcval,jcnnz,lim,lmem,flag)

      implicit none

C     SCALAR ARGUMENTS
      logical lmem
      integer flag,ind,jcnnz,lim,n

C     ARRAY ARGUMENTS
      integer jcvar(n)
      double precision x(n),jcval(n)

      flag = - 1

      end

C     ******************************************************************
C     ******************************************************************

      subroutine myevalhc(n,x,ind,hcrow,hccol,hcval,hcnnz,lim,lmem,flag)

      implicit none

C     SCALAR ARGUMENTS
      logical lmem
      integer flag,hcnnz,ind,lim,n

C     ARRAY ARGUMENTS
      integer hccol(lim),hcrow(lim)
      double precision hcval(lim),x(n)

      flag = - 1

      end

C     ******************************************************************
C     ******************************************************************

      subroutine myevalfc(n,x,f,m,c,flag)

      implicit none

C     SCALAR ARGUMENTS
      integer flag,m,n
      double precision f

C     ARRAY ARGUMENTS
      double precision c(m),x(n)

      flag = 0

C     Objective function

      f = x(2)

C     Constraints

      c(1) = x(1) ** 2 + 1 - x(n)
      c(2) = 2.0d0 - x(1) - x(n)

      end

C     ******************************************************************
C     ******************************************************************

      subroutine myevalgjac(n,x,g,m,jcfun,jcvar,jcval,jcnnz,lim,lmem,
     +flag)

      implicit none

C     SCALAR ARGUMENTS
      logical lmem
      integer flag,jcnnz,lim,m,n

C     ARRAY ARGUMENTS
      integer jcfun(lim),jcvar(lim)
      double precision g(n),jcval(lim),x(n)

      flag = 0
      lmem = .false.

C     Gradient of the objective function

      g(1) = 0.0d0
      g(2) = 1.0d0

C     Gradient of constraints

      jcnnz = 4

      if ( jcnnz .gt. lim ) then
          lmem = .true.
          return
      end if

C     Constraint 1

      jcfun(1) = 1
      jcvar(1) = 1
      jcval(1) = 2.0d0 * x(1)

      jcfun(2) = 1
      jcvar(2) = 2
      jcval(2) = - 1.0d0

C     Constraint 2

      jcfun(3) = 2
      jcvar(3) = 1
      jcval(3) = - 1.0d0

      jcfun(4) = 2
      jcvar(4) = 2
      jcval(4) = - 1.0d0

      end

C     ******************************************************************
C     ******************************************************************

      subroutine myevalgjacp(n,x,g,m,p,q,work,gotj,flag)

      implicit none

C     SCALAR ARGUMENTS
      logical gotj
      integer flag,m,n
      character work

C     ARRAY ARGUMENTS
      double precision g(n),p(m),q(n),x(n)

      flag = - 1

      end

C     ******************************************************************
C     ******************************************************************

      subroutine myevalhl(n,x,m,lambda,sf,sc,hlrow,hlcol,hlval,hlnnz,
     +lim,lmem,flag)

      implicit none

C     SCALAR ARGUMENTS
      logical lmem
      integer flag,hlnnz,lim,m,n
      double precision sf

C     ARRAY ARGUMENTS
      integer hlcol(lim),hlrow(lim)
      double precision hlval(lim),lambda(m),sc(m),x(n)

      flag = 0
      lmem = .false.

      hlnnz = 1

      if ( hlnnz .gt. lim ) then
          lmem = .true.
          return
      end if

      hlrow(1) = 1
      hlcol(1) = 1
      hlval(1) = lambda(1) * sc(1) * 2.0d0

      end

C     ******************************************************************
C     ******************************************************************

      subroutine myevalhlp(n,x,m,lambda,sf,sc,p,hp,goth,flag)

      implicit none

C     SCALAR ARGUMENTS
      logical goth
      integer flag,m,n
      double precision sf

C     ARRAY ARGUMENTS
      double precision hp(n),lambda(m),p(n),sc(m),x(n)

      flag = - 1

      end

File: algencan-3.1.1/sources/examples/c/toyprob2.c
#include <stdlib.h>
#include <stdio.h>
#include <math.h>
#include <string.h>

void c_algencan(void *myevalf, void *myevalg, void *myevalh, void *myevalc,
	void *myevaljac, void *myevalhc, void *myevalfc, void *myevalgjac,
	void *myevalgjacp, void *myevalhl, void *myevalhlp, int jcnnzmax,
       	int hnnzmax,double *epsfeas, double *epsopt, double *efstin,
	double *eostin, double *efacc, double *eoacc, char *outputfnm,
	char *specfnm, int nvparam,char **vparam, int n, double *x,
	double *l, double *u, int m, double *lambda, _Bool *equatn,
	_Bool *linear, _Bool *coded, _Bool checkder, double *f,
	double *cnorm, double *snorm, double *nlpsupn,int *inform);

/* ******************************************************************
   ****************************************************************** */

void myevalf(int n, double *x, double *f, int *flag) {

   *flag = - 1;
}

/* ******************************************************************
   ****************************************************************** */

void myevalg(int n, double *x, double *g, int *flag) {

   *flag = - 1;
}

/* ******************************************************************
   ****************************************************************** */

void myevalh(int n, double *x, int *hrow, int *hcol, double *hval, int *hnnz,
	     int lim, _Bool *lmem, int *flag) {
   *flag = - 1;
}

/* ******************************************************************
   ****************************************************************** */

void myevalc(int n, double *x, int ind, double *c, int *flag) {

   *flag = -1;
}

/* ******************************************************************
   ****************************************************************** */

void myevaljac(int n, double *x, int ind, int *jcvar, double *jcval,
	       int *jcnnz, int lim, _Bool *lmem, int *flag) {
   *flag = - 1;
}

/* ******************************************************************
   ****************************************************************** */

void myevalhc(int n, double *x, int ind, int *hcrow, int *hccol, double *hcval,
	      int *hcnnz, int lim, _Bool *lmem, int *flag) {
   *flag = - 1;
}

/* *****************************************************************
   ***************************************************************** */

void myevalfc(int n, double *x, double *f, int m, double *c, int *flag) {

  *flag = 0;

  /* Objective function */
  *f = x[1];

  /* Constraints */
  c[0] = pow( x[0], 2.0 ) + 1.0 - x[1];
  c[1] = 2.0 - x[0] - x[1];
}

/* *****************************************************************
   ***************************************************************** */

void myevalgjac(int n, double *x, double *g, int m, int *jcfun, int *jcvar,
		double *jcval, int *jcnnz, int lim, _Bool *lmem, int *flag) {

  *flag = 0;
  *lmem = 0;

  g[0] = 0.0;
  g[1] = 1.0;

  *jcnnz = 4;

  if( *jcnnz > lim ) {
    *lmem = 1;
    return;
  }

  jcfun[0] = 0;
  jcvar[0] = 0;
  jcval[0] = 2.0 * x[0];

  jcfun[1] = 0;
  jcvar[1] = 1;
  jcval[1] = - 1.0;

  jcfun[2] = 1;
  jcvar[2] = 0;
  jcval[2] = - 1.0;

  jcfun[3] = 1;
  jcvar[3] = 1;
  jcval[3] = - 1.0;
}

/* *****************************************************************
   ***************************************************************** */

void myevalgjacp(int n, double *x, double *g, int m, double *p, double *q,
		 char work, _Bool *gotj, int *flag) {

  *flag = -1;
}

/* *****************************************************************
   ***************************************************************** */

void myevalhl(int n, double *x, int m, double *lambda, double scalef,
	      double *scalec, int *hlrow, int *hlcol, double *hlval,
	      int *hlnnz, int lim, _Bool *lmem, int *flag) {

  *flag = 0;
  *lmem = 0;
  
  *hlnnz = 1;

  if( *hlnnz > lim ) {
    *lmem = 1;
    return;
  }
  
  hlrow[0] = 0;
  hlcol[0] = 0;
  hlval[0] = scalec[0] * lambda[0] * 2.0;
}

/* *****************************************************************
   ***************************************************************** */

void myevalhlp(int n, double *x, int m, double *lambda, double scalef,
	       double *scalec, double *p, double *hp, _Bool *goth, 
	       int *flag) {
  
  *flag = -1;
}

/* ******************************************************************
   ****************************************************************** */

int main() {
  _Bool  checkder;
  int    hnnzmax,hnnzmax1,hnnzmax2,hnnzmax3,i,jcnnzmax,inform,m,n,nvparam;
  double cnorm,efacc,efstin,eoacc,eostin,epsfeas,epsopt,f,nlpsupn,snorm;
  
  char   *specfnm, *outputfnm, **vparam;
  _Bool  coded[11],*equatn,*linear;
  double *l,*lambda,*u,*x;
  
  n = 2;
  m = 2;
  
  /* Memory allocation */
  x      = (double *) malloc(n * sizeof(double));
  l      = (double *) malloc(n * sizeof(double));
  u      = (double *) malloc(n * sizeof(double));
  lambda = (double *) malloc(m * sizeof(double));
  equatn = (_Bool  *) malloc(m * sizeof(_Bool ));
  linear = (_Bool  *) malloc(m * sizeof(_Bool ));
  
  if (     x == NULL ||      l == NULL ||      u == NULL ||
      lambda == NULL || equatn == NULL || linear == NULL ) {
    
    printf( "\nC INTERFACE ERROR: It was not possible to allocate memory.\n" );
    exit( 0 );
    
  }

  /* Initial point */
  for(i = 0; i < n; i++) x[i] = 0.0;
  
  /* Lower and upper bounds */
  l[0] = - 10.0;
  l[1] = - 1.0e20;
  
  u[0] = 10.0;
  u[1] = 1.0e20;
  
  /* For each constraint i, set equatn[i] = 1. if it is an equality
     constraint of the form c_i(x) = 0, and set equatn[i] = 0 if it is
     an inequality constraint of the form c_i(x) <= 0. */
  equatn[0] = 0;
  equatn[1] = 0;

  /* For each constraint i, set linear[i] = 1 if it is a linear
     constraint, otherwise set linear[i] = 0 */
  linear[0] = 0;
  linear[1] = 1;
  
  /* Lagrange multipliers approximation. */
  for( i = 0; i < m; i++ ) lambda[i] = 0.0;
  
  /* In this C interface evalf, evalg, evalh, evalc, evaljac and
     evalhc are present. evalfc, evalgjac, evalhl and evalhlp are
     not. */
  
  coded[0]  = 0; /* fsub     */
  coded[1]  = 0; /* gsub     */
  coded[2]  = 0; /* hsub     */
  coded[3]  = 0; /* csub     */
  coded[4]  = 0; /* jacsub   */
  coded[5]  = 0; /* hcsub    */
  coded[6]  = 1; /* fcsub    */
  coded[7]  = 1; /* gjacsub  */
  coded[8]  = 0; /* gjacpsub */
  coded[9]  = 1; /* hlsub    */
  coded[10] = 0; /* hlpsub   */
 
  /* Upper bounds on the number of sparse-matrices non-null
     elements */
  jcnnzmax = 4;
  hnnzmax1 = 0;
  hnnzmax2 = 1;
  hnnzmax3 = 6;
  hnnzmax  = hnnzmax1 + hnnzmax2 + hnnzmax3;

  /* Check derivatives? */
  checkder = 0;

  /* Parameters setting */
  epsfeas  = 1.0e-08;
  epsopt   = 1.0e-08;

  efstin   = sqrt( epsfeas );
  eostin   = pow( epsopt, 1.5 );

  efacc    = sqrt( epsfeas );
  eoacc    = sqrt( epsopt );

  outputfnm = "algencan.out";
  specfnm   = "algencan.dat";

  nvparam = 2;
  
  /* Allocates VPARAM array */
  vparam = ( char ** ) malloc( nvparam * sizeof( char * ) );

  /* Set algencan parameters */
  vparam[0] = "ITERATIONS-OUTPUT-DETAILS 10";
  vparam[1] = "TRUNCATED-NEWTON-LINE-SEARCH-INNER-SOLVER";

  nvparam = 0;

  /* Optimize */
  c_algencan(&myevalf,&myevalg,&myevalh,&myevalc,&myevaljac,&myevalhc,&myevalfc,
	     &myevalgjac,&myevalgjacp,&myevalhl,&myevalhlp,jcnnzmax,hnnzmax,
	     &epsfeas,&epsopt,&efstin,&eostin,&efacc,&eoacc,outputfnm,specfnm,
	     nvparam,vparam,n,x,l,u,m,lambda,equatn,linear,coded,checkder,
	     &f,&cnorm,&snorm,&nlpsupn,&inform);

  /* Memory deallocation */
  free(x     );
  free(l     );
  free(u     );
  free(lambda);
  free(equatn);
  free(linear);
}
